### PR TITLE
*: Centralize requirements.txt and unify dependency versions

### PR DESCRIPTION
**Pull request by barakmich**
## Why are these changes needed?

There are multiple places in the codebase that install pip dependencies and not all of them are in sync.

There are still some in yaml cluster files that can be fixed up, but this at least deals with our development systems.

I'm going to need a hand with running the release tests to make sure my ci changes and docker changes are fine

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: ci/travis/install-dependencies.sh [hunk 1]
```diff
@@ -275,34 +260,22 @@ install_dependencies() {
...
   # Additional Doc test dependencies.
   if [ "${DOC_TESTING-}" = 1 ]; then
-    pip install tensorflow-probability=="${TFP_VERSION-0.8}" \
-      torch=="${TORCH_VERSION-1.4}" torchvision atari_py gym[atari] lz4 smart_open
-    pip install -r "${WORKSPACE_DIR}"/docker/tune_test/requirements.txt
-  fi
-
-  # Additional streaming dependencies.
-  if [ "${RAY_CI_STREAMING_PYTHON_AFFECTED}" = 1 ]; then
-    pip install "msgpack>=1.0.0"
```
Comment (barakmich) [inline ci/travis/install-dependencies.sh:305]: this is intentional as msgpack is already a dependency of ray

File: doc/examples/cython/cython_examples/__init__.py [hunk 2]
```diff
@@ -1,3 +1,9 @@
+# flake8: noqa
+import numpy
+import pyximport
+
+pyximport.install(setup_args={"include_dirs": numpy.get_include()})
```
Comment (mehrdadn) [inline doc/examples/cython/cython_examples/__init__.py:5]: I'm guessing this should be under `if __name__ == '__main__'`?

File: doc/examples/cython/cython_examples/__init__.py [hunk 3]
```diff
@@ -11,17 +17,8 @@
...
-        "compute_single_self_corr_syrk",
-        "compute_single_self_corr_gemm",
-        "compute_corr_vectors",
-        "compute_single_matrix_multiplication"
-        ]
+    "simple_func", "fib", "fib_int", "fib_cpdef", "fib_cdef", "simple_class",
+    "masked_log", "compute_self_corr_for_voxel_sel", "compute_kernel_matrix",
+    "compute_single_self_corr_syrk", "compute_single_self_corr_gemm",
+    "compute_corr_vectors", "compute_single_matrix_multiplication"
+]
```
Comment (mehrdadn) [inline doc/examples/cython/cython_examples/__init__.py:24]: Was this a linter change or a manual one? It seems nice but has the problem of making git merges harder. Not a big deal but I would've probably kept it as is.

File: python/ray/examples/cython_examples [hunk 4]
```diff
@@ -0,0 +1 @@
+../../../doc/examples/cython/cython_examples
```
Comment (mehrdadn) [inline python/ray/examples/cython_examples:1]: Is there a way to avoid symlinks here? They cause problems for people once in a while because people edit them like text files accidentally. This happens most easily on Windows (sine Git disables symlinks by default) but I've seen it happen on non-Windows systems too (I guess people edit directly on GitHub). If we can avoid symlinks without too much trouble that'd be preferable.

File: python/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,59 @@
+# General requirements
+blist; platform_system != "Windows"
```
Comment (mehrdadn) [inline python/requirements.txt:2]: Wow this is cool! Didn't know you can do that.

File: python/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+pytest-asyncio
+pytest-rerunfailures
+pytest-sugar
+pytest-timeout
+scikit-learn==0.22.2
+tensorflow==2.2.0
+werkzeug
+xlrd
+
+# setup.py requirements
```
Comment (mehrdadn) [inline python/requirements.txt:27]: Does this need to be kept in sync with setup.py? If so it'd be nice to note that explicitly in both places (and similarly elsewhere).

File: python/setup.py [hunk 7]
```diff
@@ -310,7 +315,7 @@ def find_version(*filepath):
     "gpustat",
     "grpcio >= 1.28.1",
     "jsonschema",
-    "msgpack >= 0.6.0, < 2.0.0",
+    "msgpack >= 1.0.0, < 2.0.0",
```
Comment (mehrdadn) [inline python/setup.py:318]: Can we just replace `install_requires` with a read of the appropriate part of `requirements.txt` instead? so we don't have to keep them in sync.

File: python/setup.py [hunk 8]
```diff
@@ -114,7 +115,7 @@
     "scipy",
 ]
 
-extras["streaming"] = ["msgpack >= 0.6.2"]
+extras["streaming"] = ["msgpack >= 1.0.0"]
```
Comment (mehrdadn) [inline python/setup.py:118]: Can we factor this out into a variable or something now that it seems we have to keep it in sync?

File: python/setup.py [hunk 9]
```diff
@@ -249,7 +250,11 @@ def build(build_python, build_java):
 
     version_info = bazel_invoke(subprocess.check_output, ["--version"])
     bazel_version_str = version_info.rstrip().decode("utf-8").split(" ", 1)[1]
-    bazel_version = tuple(map(int, bazel_version_str.split(".")))
+    bazel_version_split = bazel_version_str.split(".")
+    bazel_version_digits = [
+        "".join(takewhile(str.isdigit, s)) for s in bazel_version_split
```
Comment (mehrdadn) [inline python/setup.py:255]: Nice, thanks! Which Bazel version did this break on?

**Review by mehrdadn — APPROVED** [review]
LGTM overall! Left some minor comments.

File: python/requirements_autoscaler.txt [hunk 10]
```diff
@@ -1,28 +1,20 @@
...
+keras
 lightgbm 
-xgboost 
+mlflow
+nevergrad 
+oauth2client
+scikit-optimize 
+sigopt
+smart_open
+tensorflow_probability
```
Comment (robertnishihara) [inline python/requirements_autoscaler.txt:17]: @barakmich thanks! The way we're installing dependencies right now is a bit of a mess. Can you explain the semantics of these files? Most of these dependencies are only used in the tests. So maybe this should be called `requirements_test_autoscaler.txt` or something like that? Same with RLlib, Tune, etc.

Also, maybe these should just be the dependencies that are required for the tests that are not already included in `setup.py` (instead of the union).

All dependencies that users need are listed in `setup.py`.

Basically, it'd be helpful to see a clear explanation of what the "meaning" of each requirements.txt file is and whether it is relevant for users or just testing or something else.

File: doc/examples/cython/cython_examples/__init__.py (hunk 2 above)
Comment (barakmich) [reply doc/examples/cython/cython_examples/__init__.py:5]: Nah, needs to be there so that when this is imported, it imports on the line below. So this (a) makes it possible, because it's a pyx and (b) will compile with the right include path

File: doc/examples/cython/cython_examples/__init__.py (hunk 3 above)
Comment (barakmich) [reply doc/examples/cython/cython_examples/__init__.py:24]: Yeah, this is a linter change. I also prefer the longer lines.

File: python/ray/examples/cython_examples (hunk 4 above)
Comment (barakmich) [reply python/ray/examples/cython_examples:1]: I'm not sure, as it's a documentation piece that specifically has a test in `python/ray/tests` -- either (a) we rewrite the docs to point to this example living in the main code, (b) we move the test to join the code or (c) some symlink like this. I actually prefer (b) if that's a reasonable answer to you.

File: python/setup.py (hunk 7 above)
Comment (barakmich) [reply python/setup.py:318]: I suppose I could try some minor magic here because the setup.py tends to depend on the whole repo (bazel, et al)

File: python/setup.py (hunk 8 above)
Comment (barakmich) [reply python/setup.py:118]: Actually, I can just remove this one (empty list, so ray[streaming] works) , and it'll be in one place

File: python/setup.py (hunk 9 above)
Comment (barakmich) [reply python/setup.py:255]:
```
$ bazel --version
bazel 3.4.1- (@non-git)
```
aka the one I installed from the Arch repos. Parsed wrong because of the extra bit at the end.

File: python/requirements_autoscaler.txt (hunk 10 above)
Comment (barakmich) [reply python/requirements_autoscaler.txt:17]: Yeah, that tracks. With the setup.py trick I'm about to try from the comments above, and a little guide, it should be doable.

The reason I kept setup.py separate initially is exactly that -- all the dependencies /users/ need are listed there. That's all for a standard user.

The rest are for developers of the different sub-pieces (and tests, though I didn't know that) -- requirements.txt in general being aimed at the developer-of-ray experience. So yeah, some of these can be combined for sure, if that makes sense. I just broke them out by specific use-case from the tooling (where they were repeated and often out of sync). The good news is that follow up PRs can massage these files and not have to change them in many different places.

File: python/setup.py (hunk 7 above)
Comment (barakmich) [reply python/setup.py:318]: Actually, I'll leave the bidirectional sync comments. The separation is much like Robert pointed out (user deps vs dev deps)

File: python/requirements.txt (hunk 6 above)
Comment (barakmich) [reply python/requirements.txt:27]: This is the only place that really needs syncing

File: python/ray/examples/cython_examples (hunk 4 above)
Comment (mehrdadn) [reply python/ray/examples/cython_examples:1]: I don't actually know enough to tell—you'd have to ask someone who's been involved on the code. Not a big deal either way, only worth worrying about if it's not much trouble.

File: python/setup.py (hunk 9 above)
Comment (mehrdadn) [reply python/setup.py:255]: Ahh, thanks!

File: python/setup.py (hunk 7 above)
Comment (mehrdadn) [reply python/setup.py:318]: Is there any disadvantage to reading a requirements file here? I feel like it wouldn't actually cause any issues, and it would prevent the need to keep things in sync, though I might be missing something.

Comment (barakmich) [reply python/setup.py]: Just one more file to install, and the various subpackages

File: python/ray/examples/cython_examples (hunk 4 above)
Comment (barakmich) [reply python/ray/examples/cython_examples]: yeah, it's not much trouble. and since there's a `setup.py` for the cython_examples already, I'm calling that a separate package (now with its own test)

**Comment by barakmich** [comment]
@mehrdadn @ericl -- Build is now consistent with other flaky failures in other PRs

**Comment by ericl** [comment]
Awesome, merging
